### PR TITLE
Merge branch 'master' into 3.1, excluding 3.2+ specific changes

### PR DESCRIPTION
**Pull request by akien-mga**
Merge branch 'master' into 3.1, excluding 3.2+ specific changes

I went through the diff to try to ensure that no 3.2+ specific changes
are being backported to the stable branch. I might have missed some
things though.

----

To be clear, the workflow used was:
```
git checkout 3.1
git merge master
// resolve merge conflicts
// undo all changes from `git diff --staged` which are unwanted, such as `classes/` changes
git merge --continue
```

### CODE REVIEW

**Comment by akien-mga** [comment]
The commit to review is the last one, ce1f84f5a, which should show all the *actual* changes that will be in `3.1/HEAD` once merged.

File: getting_started/step_by_step/signals.rst [hunk 1]
```diff
@@ -199,126 +204,6 @@ To emit a signal via code, use the ``emit`` function:
         }
     }
 
-Shooting example
```
Comment (akien-mga) [inline getting_started/step_by_step/signals.rst]: To be clear, this is not removed, but moved to its own page.

File: tutorials/2d/2d_sprite_animation.rst [hunk 2]
```diff
@@ -0,0 +1,155 @@
+.. _doc_2d_sprite_animation:
+
+2D Sprite animation
```
Comment (akien-mga) [inline tutorials/2d/2d_sprite_animation.rst]: AFAIK this tutorial is compatible with 3.1 (i.e. does not cover the new Atlas packaging feature in 3.2). CC @cbscribe to confirm.

File: tutorials/2d/using_tilemaps.rst [hunk 3]
```diff
@@ -6,178 +6,162 @@ Using tilemaps
 Introduction
 ------------
 
-Tilemaps are a simple and quick way to make 2D game levels. Basically,
-you start with a bunch of reference tiles (or pieces) that can be put on a
-grid, as many times each as desired - think of it like a map editor:
-
-.. image:: img/tilemap.png
+A tilemap is a grid of tiles used to create a game's layout. There are several
```
Comment (akien-mga) [inline tutorials/2d/using_tilemaps.rst]: This was updated in 2c645c823 for 3.1, so I assume it's good for 3.1.1-stable (CC @cbscribe).

File: tutorials/animation/cutout_animation.rst [hunk 4]
```diff
@@ -6,47 +6,46 @@ Cutout animation
...
-`here <https://en.wikipedia.org/wiki/Cutout_animation>`__).
-
-With the advent of the digital age, this technique became possible using
-computers, which resulted in an increased amount of animation TV shows
-using digital Cut-out. Notable examples are `South
-Park <https://en.wikipedia.org/wiki/South_Park>`__ or `Jake and the Never
-Land
-Pirates <https://en.wikipedia.org/wiki/Jake_and_the_Never_Land_Pirates>`__
-.
+Traditionally, `cutout animation <https://en.wikipedia.org/wiki/Cutout_animation>`__
```
Comment (akien-mga) [inline tutorials/animation/cutout_animation.rst]: This was an edit "for clarity" in d12a5c81, so I assume it does not introduce 3.2-specific concepts.

File: tutorials/audio/audio_buses.rst [hunk 5]
```diff
@@ -6,34 +6,23 @@ Audio buses
 Introduction
 ------------
 
-Beginning with Godot 3.0, the audio engine has been rewritten from scratch.
-The aim now is to present an interface much friendlier to sound design
-professionals. To achieve this, the audio engine contains a virtual rack
-where unlimited audio buses can be created and, on each of it, unlimited
-amount of effect processors can be added (or more like, as long as your
-CPU holds up!)
+Godot's audio processing code has been written with games in mind, with the aim of achieving an optimal balance between performance and sound quality.
```
Comment (akien-mga) [inline tutorials/audio/audio_buses.rst]: There were some audio changes in master, but AFAIK this page does not include them. Those changes are from a6ef7af39, and while the commit log is not very descriptive, it looks like a language/clarity copyedit.

File: tutorials/audio/audio_streams.rst [hunk 6]
```diff
@@ -6,43 +6,41 @@ Audio streams
 Introduction
 ------------
 
-As you might have already read in the :ref:`Audio Buses Tutorial<doc_audio-buses>` ,
+As you might have already read in the :ref:`Audio Buses Tutorial<doc_audio-buses>`,
```
Comment (akien-mga) [inline tutorials/audio/audio_streams.rst]: Also a copyedit: ab4b468.

File: tutorials/content/procedural_geometry.rst [hunk 7]
```diff
@@ -0,0 +1,144 @@
+.. _doc_procedural_geometry:
```
Comment (akien-mga) [inline tutorials/content/procedural_geometry.rst]: This was added by reduz in 131c90b6, so needs checking to ensure it works in 3.1.x too.

File: tutorials/optimization/using_servers.rst [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+  method will return the viewport RID in the server.
+* For 3D, the :ref:`World <class_World>` resource (obtainable in the :ref:`Viewport <class_Viewport>`
+  and :ref:`Spatial <class_Spatial>` nodes)
+  contains functions to get the *VisualServer Scenario*, and the *PhysicsServer Space*. This
+  allows creating 3D objects directly with the server API and using them.
+* For 2D, the :ref:`World2D <class_World2D>` resource (obtainable in the :ref:`Viewport <class_Viewport>`
+  and :ref:`CanvasItem <class_CanvasItem>` nodes)
+  contains functions to get the *VisualServer Canvas*, and the *Physics2DServer Space*. This
+  allows creating 2D objects directly with the server API and using them.
+* The :ref:`VisualInstance<class_VisualInstance>` class, allows getting the scenario *instance* and
```
Comment (akien-mga) [inline tutorials/optimization/using_servers.rst]: Those two methods were added in the master branch, but I cherrypicked them in 3.1.1-stable, so they should work fine in the current stable.

Might be worth adding a note that they're only in 3.1.1+ though.

File: tutorials/optimization/using_multimesh.rst [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+to the shader can be provided via textures (there are floating point :ref:`Image<class_Image>` formats
+which are ideal for this).
+
+Another alternative is to use GDNative and C++, which should be extremely efficient (it's possible
+to set the entire state for all objects using linear memory via the
+:ref:`VisualServer.multimesh_set_as_bulk_array() <class_VisualServer_method_multimesh_set_as_bulk_array>`
+function). This way, the array can be created with multiple threads, then set in one call, providing
+high cache efficiency.
+
+Finally, it's not required to have all MultiMesh instances visible. The amount of visible ones can be
```
Comment (akien-mga) [inline tutorials/optimization/using_multimesh.rst]: This method was also added in master and cherrypicked for 3.1.1-stable, I'll add a note to mention it.

File: tutorials/content/procedural_geometry.rst (hunk 7 above)
Comment (akien-mga) [reply tutorials/content/procedural_geometry.rst]: Had a quick look, all the methods used here seem to be bound in `3.1-stable`.

File: tutorials/gui/gui_containers.rst [hunk 10]
```diff
@@ -0,0 +1,181 @@
+.. _doc_gui_containers:
```
Comment (akien-mga) [inline tutorials/gui/gui_containers.rst]: Added by reduz in dbf5a855d5, from a quick look it doesn't seem to have 3.2-only stuff.

File: tutorials/shading/your_first_shader/index.rst [hunk 11]
```diff
@@ -0,0 +1,26 @@
+Your first shader
```
Comment (akien-mga) [inline tutorials/shading/your_first_shader/index.rst]: I did not check whether all the new "Your first shader" tutorials are compatible with 3.1.x, CC @clayjohn.

**Comment by akien-mga** [comment]
I've done a relatively thorough review and @NathanLovato also checked the diff quickly, so let's merge.

File: tutorials/shading/your_first_shader/index.rst (hunk 11 above)
Comment (clayjohn) [reply tutorials/shading/your_first_shader/index.rst]: Yep! These were written in 3.1 So its all good!

**Pull request merged**